core/src/apps/cardano/address.py

```diff
@@ -1,145 +1,344 @@
-    if not isinstance(address_data_encoded, bytes):
+def _validate_output_shelley_address(
+    address_str: str, address_bytes: bytes, network_id: int
+) -> None:
+    address_type = _get_address_type(address_bytes)
+    if address_type == CardanoAddressType.REWARD:
         raise INVALID_ADDRESS
 
-    address_crc = address_unpacked[1]
-    if not isinstance(address_crc, int):
-        raise INVALID_ADDRESS
+    _validate_output_address_bech32_hrp(address_str, address_type, network_id)
+    _validate_address_network_id(address_bytes, network_id)
+
 
-    if address_crc != crc.crc32(address_data_encoded):
+def _validate_output_address_bech32_hrp(
+    address_str: str, address_type: int, network_id: int
+) -> None:
+    valid_hrp = _get_bech32_hrp_for_address(address_type, network_id)
+    bech32_hrp = bech32.get_hrp(address_str)
+
+    if valid_hrp != bech32_hrp:
         raise INVALID_ADDRESS
 
-    return address_data_encoded
 
+def _get_bech32_hrp_for_address(address_type: int, network_id: int) -> str:
+    if address_type == CardanoAddressType.BYRON:
+        raise ValueError("Byron address uses base58 encoding")
 
-def _validate_address_data_protocol_magic(
-    address_data_encoded: bytes, protocol_magic: int
-) -> None:
-    """
-    Determines whether the correct protocol magic (or none)
-    is included in the address. Addresses on mainnet don't
-    contain protocol magic, but addresses on the testnet do.
-    """
-    address_data = cbor.decode(address_data_encoded)
-    if not isinstance(address_data, list) or len(address_data) < 2:
-        raise INVALID_ADDRESS
+    if address_type == CardanoAddressType.REWARD:
+        if network_ids.is_mainnet(network_id):
+            return bech32.HRP_REWARD_ADDRESS
+        else:
+            return bech32.HRP_TESTNET_REWARD_ADDRESS
+    else:
+        if network_ids.is_mainnet(network_id):
+            return bech32.HRP_ADDRESS
+        else:
+            return bech32.HRP_TESTNET_ADDRESS
 
-    attributes = address_data[1]
-    if protocol_magic == protocol_magics.MAINNET:
-        if PROTOCOL_MAGIC_KEY in attributes:
-            raise NETWORK_MISMATCH
-    else:  # testnet
-        if len(attributes) == 0 or PROTOCOL_MAGIC_KEY not in attributes:
-            raise NETWORK_MISMATCH
 
-        protocol_magic_cbor = attributes[PROTOCOL_MAGIC_KEY]
-        address_protocol_magic = cbor.decode(protocol_magic_cbor)
+def _validate_address_network_id(address: bytes, network_id: int) -> None:
+    if _get_address_network_id(address) != network_id:
+        raise NETWORK_MISMATCH
+
+
+def _get_address_network_id(address: bytes) -> int:
+    return address[0] & 0x0F
+
+
+def get_public_key_hash(keychain: seed.Keychain, path: List[int]) -> bytes:
+    node = keychain.derive(path)
+    public_key = remove_ed25519_prefix(node.public_key())
+    return hashlib.blake2b(data=public_key, outlen=28).digest()
+
+
+def get_human_readable_address(address: bytes, network_id: int) -> str:
```

Comment:
I guess this function was used somewhere else before some refactoring. Since it isn't anymore, it doesn't need to be separate. Removed [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/76bc4dcc3a298bc190d0cd4263fae7ed1d998d73).